### PR TITLE
Add: encode unit test

### PR DESCRIPTION
**Pull request by gaoxinge**
**What this PR does**:

Add encode unit test for binary, date, null, object, string, map.

**Does this PR introduce a user-facing change?**:

```release-note
none
```

### CODE REVIEW

File: encode_test.go [hunk 1]
```diff
@@ -45,14 +45,10 @@ func javaDecodeValidate(method string, target interface{}) (string, error) {
...
-	_, e = stdin.Write(b)
-	if e != nil {
-		return "", e
-	}
-	e = stdin.Close()
-	if e != nil {
-		return "", e
-	}
+	go func() {
+		stdin.Write(b)
```
Comment (wongoo) [inline encode_test.go]: why run in new goroutine?

File: binary.go [hunk 2]
```diff
@@ -38,8 +38,7 @@ func encBinary(b []byte, v []byte) []byte {
 	)
 
 	if len(v) == 0 {
-		//return encByte(b, BC_BINARY_DIRECT)
-		return encByte(b, BC_NULL)
+		return encByte(b, BC_BINARY_DIRECT)
```
Comment (wongoo) [inline binary.go]: java encode as null for null binary:

```
    byte []data = (byte []) obj;
    
    if (data != null)
      out.writeBytes(data, 0, data.length);
    else
      out.writeNull();
```

File: binary_test.go [hunk 3]
```diff
@@ -124,3 +124,27 @@ func TestBinary(t *testing.T) {
...
+	for i := 0; i < 16; i++ {
+		s1024 += fmt.Sprintf("%02d 456789012345678901234567890123456789012345678901234567890123\n", i)
+	}
+
+	s65560 := ""
+	for i := 0; i < 1024; i++ {
+		s65560 += fmt.Sprintf("%03d 56789012345678901234567890123456789012345678901234567890123\n", i)
+	}
+
+	testJavaDecode(t, "argBinary_0", []byte(s0))
```
Comment (wongoo) [inline binary_test.go]: `argBinary_0` response binary array with one zero byte:
```
  public Object replyBinary_0()
  {
    return new byte[0];
  }
```

File: object_test.go [hunk 4]
```diff
@@ -312,7 +312,7 @@ func (*A16) JavaClassName() string {
 }
 
 type TestObjectStruct struct {
-	Value int `hessian:"_value"`
+	Value int32 `hessian:"_value"`
```
Comment (wongoo) [inline object_test.go]: why changed to int32?

File: encode_test.go (hunk 1 above)
Comment (gaoxinge) [reply encode_test.go]: Firstly `StdinPipe` has byte number limit. So if a process fills the pipe fully by writing bytes and no process read bytes from the pipe, then the writing process will hang. One can find explicit limit number [here](https://stackoverflow.com/a/50118778/5772365).

Origin code is below:

https://github.com/dubbogo/hessian2/blob/a7ee7ab1292618d576895782e91fb8491ef50e49/encode_test.go#L45-L57

We can see that the writing process is the go program and the reading process is the java program. Further more go program starts writing from `line 48` and java program starts reading from `line 57`.

So if byte's number is larger than 4096 on windows, i.e. like [here](https://github.com/dubbogo/hessian2/pull/68/files#diff-31acfea64559ab28bbb17d31d8e00de0R142), go program will hang on write, and the java program will never start reading.

So we should run a writing goroutine to avoid hanging the reading process.

File: binary.go (hunk 2 above)
Comment (wongoo) [reply binary.go]: it should encode BC_NULL if v is nil

Comment (gaoxinge) [reply binary.go]: Yes. I have modified the code.

File: binary_test.go (hunk 3 above)
Comment (gaoxinge) [reply binary_test.go]: En, I may not understand what's your meaning. Because `s0 := ""`,  I think `[]byte(s0)` in go equals `new byte[0]` in java.

File: object_test.go (hunk 4 above)
Comment (gaoxinge) [reply object_test.go]:
```java
public Object replyObject_1() {
    return new TestObject(Integer.valueOf(0));
}
```
Java code use the `Interger` to represent `0`, so `0` will be encode as int32.

However in go `int` type is encode as int64:

https://github.com/dubbogo/hessian2/blob/adaf25425872fad0d85c75a7a81e492f5f39f7e5/encode.go#L86-L94
